test/FunctionalTestUtils/TelemetryTestsBase.cs

```diff
@@ -23,6 +23,28 @@ public abstract class TelemetryTestsBase
         protected const int TestTimeoutMs = 10000;
         private object noParallelism = new object();
 
+        /// <summary>
+        /// Ideally, when the WebHost sever get disposed, all requests will be finished.
+        /// However, the current behavior is server dispose doesn't wait for the requests to be finished.
+        /// And during server disposing, the injected AI service will also get disposed, which means some telemetry might not be sent after server get disposed.
+        /// So we add this helper funciton to make sure all telemtry are sent before checking anything.
+        /// </summary>
+        /// <param name="condition">The condition to meet</param>
+        /// <param name="timeoutMs">Timeout in ms</param>
+        /// <param name="intervalMs">Sleep interval in ms</param>
+        public static void ConditionalTimeout(Func<bool> condition, int timeoutMs = 10000, int intervalMs = 50)
+        {
+            DateTime timeout = DateTime.Now.AddMilliseconds(timeoutMs);
+            while (DateTime.Now < timeout)
+            {
+                if (condition())
+                {
+                    return;
+                }
+                Thread.Sleep(intervalMs);
+            }
```

Comment:
Suggest: ```throw new TimeoutException();```
Or, have this method return a ```bool``` (true=>condition met; false=>timeout) and ```Assert.IsTrue``` in the caller.